support-workers/src/main/scala/com/gu/emailservices/PaperFieldsGenerator.scala

```diff
@@ -1,51 +1,70 @@
 package com.gu.emailservices
 
-import com.gu.i18n.Currency
-import com.gu.support.promotions.Promotion
+import com.gu.i18n.{Country, Currency}
+import com.gu.support.catalog.ProductRatePlanId
+import com.gu.support.promotions.{PromoCode, Promotion, PromotionService}
 import com.gu.support.workers._
-import com.gu.support.workers.states.PaymentMethodWithSchedule
+import com.gu.support.workers.states.PurchaseInfo
 import org.joda.time.LocalDate
 
-object PaperFieldsGenerator {
+import scala.concurrent.{ExecutionContext, Future}
+
+class PaperFieldsGenerator(
+  promotionService: PromotionService,
+  getMandate: String => Future[Option[String]],
+) {
+
+  def getAppliedPromotion(
+    maybePromoCode: Option[PromoCode],
+    country: Country,
+    productRatePlanId: ProductRatePlanId
+  ): Option[Promotion] =
+    for {
+      promoCode <- maybePromoCode
+      promotionWithCode <- promotionService.findPromotion(promoCode).toOption
+      validPromotion <- promotionService.validatePromotion(promotionWithCode, country, productRatePlanId, isRenewal = false).toOption
+    } yield validPromotion.promotion
 
   def fieldsFor(
-    subscriptionNumber: String,
-    billingPeriod: BillingPeriod,
+    purchaseInfo: PurchaseInfo,
+    product: ProductType,
     user: User,
-    firstDeliveryDate: Option[LocalDate],
-    currency: Currency,
-    paymentMethodWithSchedule: PaymentMethodWithSchedule,
-    directDebitMandateId: Option[String],
-    promotion: Option[Promotion],
-    giftRecipient: Option[GiftRecipient]
-  ): List[(String, String)] = {
+    productRatePlanId: Option[ProductRatePlanId],
+    fixedTerm: Boolean,
+    firstDeliveryDate: LocalDate
```

Comment:
ideally the structure of the Fields Generators should generally reflect the structure of the new types.  Given this code is special to paper+GW there shouldn't really be still 6 parameters to this method, everything should have arrived in one class. since this is a bit lower level this maybe a todo for another PR, depending how the test changes look.